Add JSON format logging support

CatherineSue opened the pull request:
JSON format log can be useful for logging infrastructure analyse.
loguru is pretty nicely integrated with fastAPI. It provides flexible logging format and retention, comparing to python-json-logger and structlog.

- Use loguru for json format log output
- Add `--logging-config` argument in openai server entrypoint to set up the logging

Closes #2397

CatherineSue committed:
Add json log format support

CatherineSue committed:
Add option to setup logging config from entrypoint

CatherineSue committed:
Format quote

simon-mo commented on vllm/logging_config.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "logger": {
+    "structured_log_file_path": "/logs/vllm/application_structured.log",
+    "unstructured_log_file_path": "/logs/vllm/application.log",
```
for default, let's consider logging to /tmp

NikolaBorisov replied:
I think by default there should not be any files written. vllm doesn't write files right now, so it I think this should stay the same. Whoever wants to have log files could pass some config.  Writing files could slow down the application. If you decide to write files by default the right place is /var/log/vllm.log

CatherineSue replied:
Shall I move it to /var or /tmp?
We can consider set the default behavior to no-file, no-json. I have added the option in the latest logging_config.json.

simon-mo reviewed (commented):
Thank you for your contribution. I believe this is a great step forward! few more questions...

NikolaBorisov reviewed (commented):
Just wanted to say, that I don't think this PR adds value to majority of vllm users, but it adds extra dependencies and complexity. There are other options to extract events from vllm like OpenTelemetry that might make more sense. Just my feeling.

simon-mo commented:
JSON logging itself (and the ability to write to files as well) is quite useful for production use cases (so you can extract the request content, also run analytics later). Additionally, this PR doesn't add complexity for the critical path. Regular output is still un-structured. I will make sure the existing workflow isn't impacted.

NikolaBorisov commented:
> JSON logging itself (and the ability to write to files as well) is quite useful for production use cases (so you can extract the request content, also run analytics later). Additionally, this PR doesn't add complexity for the critical path. Regular output is still un-structured. I will make sure the existing workflow isn't impacted.

Sure, I  just think the default should be no-files, no-json, just stderr like now. Additionally using external libraries for logging has risk. I remember having to deal with this: https://en.wikipedia.org/wiki/Log4Shell. Anyway, it is probably fine ;)

CatherineSue commented on vllm/logger.py:
```diff
@@ -1,56 +1,153 @@
-# Adapted from
-# https://github.com/skypilot-org/skypilot/blob/86dc0f6283a335e4aa37b3c10716f90999f48ab6/sky/sky_logging.py
+# Copyright © [2023,] 2023, Oracle and/or its affiliates.
```
This line is a requirement from my org.

njhill replied:
What about the preexisting lines in the file from other contributors, or when changes/additions are inevitably made by other contributors in future?

[Recommended practice](https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects#:~:text=Copyright%20Notices%20%E2%80%93%20Community%20Best%20Practice&text=Instead%2C%20many%20LF%20project%20communities,Contributors%20to%20the%20XYZ%20project.) is to include a generic copyright notice, attributing to e.g. "The vLLM authors".

CatherineSue replied:
Removed this line.

CatherineSue commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -111,11 +107,21 @@ def parse_args():
 
+args = parse_args()
+setup_logger(args.logging_config)
+logger = init_logger(__name__)
```
Yes, thanks for the suggestion, moved.

CatherineSue commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -111,11 +107,21 @@ def parse_args():
+setup_logger(args.logging_config)
+logger = init_logger(__name__)
+
+app = fastapi.FastAPI(lifespan=lifespan)
```
No, reverted.

CatherineSue commented on vllm/logger.py:
```diff
@@ -1,56 +1,153 @@
 
-class NewLineFormatter(logging.Formatter):
-    """Adds logging prefix to newlines to align multi-line messages."""
+class InterceptHandler(logging.Handler):
```
This class is to manage log messages generated by Python's standard logging library, as some third-party libraries might utilize this module.

CatherineSue commented on vllm/logger.py:
```diff
@@ -1,56 +1,153 @@
-_root_logger = logging.getLogger("vllm")
-_default_handler = None
 
+class CustomizeLogger:
```
Done

CatherineSue commented:
Sorry for the late updates. I was on vacation last week.

CatherineSue commented:
Rebase to main

CatherineSue commented:
Rebase to main

CatherineSue commented:
@simon-mo All comments have been addressed. Can you take another look when you have time? Thanks

Yard1 commented:
Could we make sure to respect the `VLLM_CONFIGURE_LOGGING` env var? vLLM should not change any logging formatting if it's set.

CatherineSue committed:
Resolve comments

- Fix CI
- Add documentation
- Make uvicorn use customized logging

CatherineSue committed:
Respect VLLM_CONFIGURE_LOGGING

CatherineSue commented:
@Yard1 Rebased on `VLLM_CONFIGURE_LOGGING`

CatherineSue committed:
Force add default_logging_config.json

CatherineSue commented:
@Yard1 I wonder if `VLLM_CONFIGURE_LOGGING` is necessary given the goal is to make it easier to integrate to some specific logging formats, then they can simply pass a new `log-config.json`

Yard1 commented:
@CatherineSue I think the env var is nice because if someone has already configured the python logging in the current process (eg. when vLLM is integrated in an external library), then there is no extra work required to keep consistent format across the entire library.

CatherineSue committed:
Fix json log exception

Context
The current json log formatter does not fully traverse traceback

- Add detailed traceback to exception logs by traversing back to
`__cause__` and `__context__` exceptions

CatherineSue committed:
Fix AMD-Test

- Need to add loguru in requirements-rocm.txt

CatherineSue committed:
Change logger.warn to logger.warning in custom_all_reduce.py

- logger.warn has been deprecated since Python3.3
- loguru doesn't support logger.warn

CatherineSue committed:
Remove Copyright line

joerunde commented on requirements-rocm.txt:
```diff
@@ -13,3 +13,4 @@ uvicorn[standard]
 pydantic >= 2.0  # Required for OpenAI server.
 prometheus_client >= 0.18.0
 outlines == 0.0.34
+loguru
```
```suggestion
loguru==0.7.2
```

Can we pin this to avoid breaking builds later when loguru updates?

joerunde commented on requirements.txt:
```diff
@@ -15,3 +15,4 @@ pynvml == 11.5.0
 triton >= 2.1.0
 outlines == 0.0.34
 tiktoken == 0.6.0 # Required for DBRX tokenizer
+loguru
```
```suggestion
loguru == 0.7.2
```

joerunde commented on vllm/logger.py:
```diff
@@ -1,61 +1,255 @@
+        config = cls.load_logging_config(config_path)
+        logging_config = config.get("logger")
+
+        logger = cls.customize_logging(
```
I think this will shadow `from loguru import logger`

Can we rename this to `_logger`? or `logger_`? `logger_instance`?

rkooo567 replied:
+1 also add a type annotation here?

joerunde commented on vllm/logger.py:
```diff
@@ -1,61 +1,255 @@
+    """
+
+    @classmethod
+    def make_logger(cls, config_path: Path):
```
It might be nice to add return annotations to the methods in this class as well:
```suggestion
def make_logger(cls, config_path: Path) -> loguru.Logger:
```

I see the contributing guidelines say 
>  Note: Currently, the repository does not pass the mypy tests.

and I think this would be required to make mypy happy, though I see a large portion of this repo does not have return type annotations :(

joerunde reviewed (commented):
I just came across this PR while looking at some other log changes, this looks excellent and I'd love to see it merged in. Thanks @CatherineSue !

rkooo567 commented on vllm/default_logging_config.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "logger": {
+    "structured_log_file_path": "/var/log/vllm/application_structured.log",
```
Feel like writing to a file by default needs a lot more work than this. There are some problems here;

1. /var is not accessible in many environments. For example, I tried this in my env and failed immediately. I recommend you to use `/tmp/vllm`. It's been done this way from ray https://github.com/ray-project/ray/, and it worked pretty well.
2. If you start a new vllm instance, what's going to happen to the previous logs? Is it overwritten? Ideally we should create a directory for each new instance

I'd suggest to not write to a file by default (same as master) and add an example to the doc instead.

rkooo567 commented on vllm/logger.py:
```diff
@@ -1,61 +1,255 @@
-    def __init__(self, fmt, datefmt=None):
-        logging.Formatter.__init__(self, fmt, datefmt)
+        # Basic configuration for intercepting standard logging messages
+        logging.basicConfig(handlers=[InterceptHandler()], level=0)
```
Does this screw up basic config from users? I think we should not touch the root config because vllm can be just imported and used as a library with other user code.

rkooo567 replied:
Can we just setup the logger for "vllm" logger and emit to stdout from there instead?

rkooo567 commented on vllm/logger.py:
```diff
@@ -1,61 +1,255 @@
-    _setup_logger()
+    _root_logger = CustomizeLogger.make_logger(config_path)
+else:
+    _root_logger = logging.getLogger("vllm")
```
IIUC, if VLLM_CONFIGURE_LOGGING is False, doesn't this have regression? I.e., don't we need to do 

```
def _setup_logger():
    _root_logger.setLevel(logging.DEBUG)
    global _default_handler
    if _default_handler is None:
        _default_handler = logging.StreamHandler(sys.stdout)
        _default_handler.flush = sys.stdout.flush  # type: ignore
        _default_handler.setLevel(logging.INFO)
        _root_logger.addHandler(_default_handler)
    fmt = NewLineFormatter(_FORMAT, datefmt=_DATE_FORMAT)
    _default_handler.setFormatter(fmt)
    # Setting this will avoid the message
    # being propagated to the parent logger.
    _root_logger.propagate = False
```
?

rkooo567 commented on vllm/logger.py:
```diff
@@ -1,61 +1,255 @@
+            level=logging_config.get("level"),
+            retention=logging_config.get("retention"),
+            rotation=logging_config.get("rotation"),
+            format=logging_config.get("format"),
```
Can we just do 

```
customize_logging(**logging_config)
```
?

rkooo567 commented on vllm/logger.py:
```diff
@@ -1,61 +1,255 @@
+            depth += 1
+
+        logger.opt(depth=depth,
+                   exception=record.exc_info).log(level, record.getMessage())
```
QQ: is `exception` arg necessary? Based on https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.opt, it says it is automatically resolved by `sys.exc_info()`?

```
exception ([bool](https://docs.python.org/3/library/functions.html#bool), [tuple](https://docs.python.org/3/library/stdtypes.html#tuple) or [Exception](https://docs.python.org/3/library/exceptions.html#Exception), optional) – If it does not evaluate as False, the passed exception is formatted and added to the log message. It could be an [Exception](https://docs.python.org/3/library/exceptions.html#Exception) object or a (type, value, traceback) tuple, otherwise the exception information is retrieved from [sys.exc_info()](https://docs.python.org/3/library/sys.html#sys.exc_info).
```

rkooo567 reviewed (commented):
I think there are some major comments to be addressed.

1. Doc needs to be updated (with examples).
2. The default seems a bit fragile. Can we follow up here? 
3. Is it possible to make it off by default first? This way, we can make sure existing feature is not regressed and we can turn it on by default as a follow up

simon-mo commented:
Hi @CatherineSue, once the rounds of comments from @rkooo567 are addressed I think this PR should be in a good shape to merge. Sorry about the delay.

CatherineSue commented:
@rkooo567 @simon-mo thank you for the comments. Will start working on them soon. ETA early next week.

rkooo567 commented:
Sounds great @CatherineSue also feel free to lmk if you want to discuss anything in person!

CatherineSue commented:
@rkooo567 Apologize for the delay. It's been very busy this week and I have been focusing on the embedding PR. Will see if I have bandwidth for this next week.

rkooo567 commented:
No problems! Very excited to land this PR!

icco commented:
Any update here? Would love to see this landed.

haitwang-cloud commented:
/ready

hmellor commented:
@CatherineSue do you have the time to update this PR?

hmellor commented:
Closing as superseded by https://github.com/vllm-project/vllm/pull/13920
